src/rt/cpp/cown.h

```diff
@@ -0,0 +1,193 @@
+// Copyright Microsoft and Project Verona Contributors.
+// SPDX-License-Identifier: MIT
+
+#include <functional>
+#include <tuple>
+#include <utility>
+#include <verona.h>
+
+/**
+ * Used in static asserts to check passed values are cown_ptr types.
+ *
+ * This class is used so a single simple check can be used
+ *   std::is_base_of<cown_ptr_base, T>
+ * To check T is a cown_ptr.
+ */
+class cown_ptr_base
+{
+private:
+  cown_ptr_base() {}
+
+  /**
+   * Only cown_ptr can construct one of these,
+   * so anything that has this base class will be a cown_ptr.
+   */
+  template<typename T>
+  friend class cown_ptr;
+};
+
+/**
+ * Smart pointer to represent shared access to a cown.
+ * Can only be used asychronously with `when` to get
+ * underlying access.
+ * 
+ * Note using lower case name to match C++ std library
+ * as this is one of the exposed types.
+ */
+template<typename T>
+class cown_ptr : cown_ptr_base
+{
+private:
+  /**
+   * Internal Verona runtime cown for this type.
+   */
+  class ActualCown : public VCown<ActualCown>
+  {
+  private:
+    T value;
+
+  public:
+    template<typename... Args>
+    ActualCown(Args&&... ts) : value(std::forward<Args>(ts)...) {}
+
+    template<typename TT>
+    friend class acquired_cown;
+  };
+
+  /**
+   * Internal Verona runtime cown for this type.
+   */
+  ActualCown* allocated_cown = nullptr;
+
+  /**
+   * Accesses the internal Verona runtime cown for this handle.
+   */
+  Cown* underlying_cown()
+  {
+    return allocated_cown;
+  }
+
+  /**
+   * Construct a new cown ptr object, actually allocates a runtime cown.
+   *
+   * This is internal, and the `make_cown` below is the public interface,
+   * which has better behaviour for implicit template arguments.
+   */
+  cown_ptr(ActualCown* cown) : allocated_cown(cown) {}
+
+public:
+  /**
+   * Copy an existing cown ptr.  Shares the underlying cown.
+   */
+  cown_ptr(const cown_ptr& other)
+  {
+    allocated_cown = other.allocated_cown;
+    verona::rt::Cown::acquire(allocated_cown);
+  }
+
+  /**
+   * Move an existing cown ptr.  Does not create a new cown,
+   * and is more efficient than copying, as it does not need
+   * to perform reference count operations.
+   */
+  cown_ptr(cown_ptr&& other)
+  {
+    allocated_cown = other.allocated_cown;
+    other.allocated_cown = nullptr;
+  }
+
+  ~cown_ptr()
+  {
+    // Condition to handle moved cown ptrs.
+    if (allocated_cown != nullptr)
+    {
+      auto& alloc = verona::rt::ThreadAlloc::get();
+      verona::rt::Cown::release(alloc, allocated_cown);

```

Comment:
In the copy ctor above, does this release after the new copy acquire it? As is this the intended semantics?

Reply:
It needs to be like this.  If you want to optimise, the caller can use the move constructor.